advanced_source/cpp_frontend.rst

```diff
@@ -0,0 +1,1346 @@
+
+.. code-block:: cpp
+
+  torch::Device device = torch::kCPU;
+  if (torch::cuda::is_available()) {
+    std::cout << "CUDA is available! Training on GPU." << std::endl;
+    device = torch::kCUDA;
+  }
+
+or even
+
+.. code-block:: cpp
+
+  torch::Device device(torch::cuda::is_available() ? torch::kCUDA : torch::kCPU);
+
+Checkpointing and Recovering the Training State
+-----------------------------------------------
+
+The last augmentation we should make to our training script is to periodically
+save the state of our model parameters, the state of our optimizers as well as a
+few generated image samples. If our computer were to crash in the middle of the
+training procedure, the first two will allow us to restore the training state.
+For long-lasting training sessions, this is absolutely essential. Fortunately,
+the C++ frontend provides an API to serialize and deserialize both model and
+optimizer state, as well as individual tensors.
+
+The core API for this is ``torch::save(thing,filename)`` and
+``torch::load(thing,filename)``, where ``thing`` could be a
+``torch::nn::Module`` subclass or an optimizer instance like the ``Adam`` object
+we have in our training script. Let's update our training loop to checkpoint the
+model and optimizer state at a certain interval:
+
+.. code-block:: cpp
+
+  if (batch_index % kCheckpointEvery == 0) {
+    // Checkpoint the model and optimizer state.
+    torch::save(generator, "generator-checkpoint.pt");
+    torch::save(generator_optimizer, "generator-optimizer-checkpoint.pt");
+    torch::save(discriminator, "discriminator-checkpoint.pt");
+    torch::save(discriminator_optimizer, "discriminator-optimizer-checkpoint.pt");
+    // Sample the generator and save the images.
+    torch::Tensor samples = generator->forward(torch::randn({8, kNoiseSize, 1, 1}, device));
+    torch::save((samples + 1.0) / 2.0, torch::str("dcgan-sample-", checkpoint_counter, ".pt"));
+    std::cout << "\n-> checkpoint " << ++checkpoint_counter << '\n';
+  }
+
+where ``kCheckpointEvery`` is an integer set to something like ``100`` to
+checkpoint every ``100`` batches, and ``checkpoint_counter`` is a counter bumped
+every time we make a checkpoint.
+
+To restore the training state, you can add lines like these after all models and
+optimizers are created, but before the training loop:
+
+.. code-block:: cpp
+
+  torch::optim::Adam generator_optimizer(
+      generator->parameters(), torch::optim::AdamOptions(2e-4).beta1(0.5));
+  torch::optim::Adam discriminator_optimizer(
+      discriminator->parameters(), torch::optim::AdamOptions(2e-4).beta1(0.5));
+
+  if (kRestoreFromCheckpoint) {
+    torch::load(generator, "generator-checkpoint.pt");
+    torch::load(generator_optimizer, "generator-optimizer-checkpoint.pt");
+    torch::load(discriminator, "discriminator-checkpoint.pt");
+    torch::load(
+        discriminator_optimizer, "discriminator-optimizer-checkpoint.pt");
```

Comment:
Just curious: how does C++ API deal with optimizers holding references to module parameters in this case? In Python one has to be careful to pickle the optimizer together with a module to keep the sharing intact, but in here you're not doing that.

Reply:
Tensor deserialization is in-place in the C++ serialization API, so the tensors to which the optimizers hold references should stay valid (as I see it)